File: BibliotecaUFABC/app/src/main/java/com/nintersoft/bibliotecaufabc/webviewclients/SyncWebClient.java
package com.nintersoft.bibliotecaufabc.webviewclients;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.Build;
import android.preference.PreferenceManager;
import android.webkit.ValueCallback;
import android.webkit.WebResourceError;
import android.webkit.WebResourceRequest;
import android.webkit.WebView;
import android.webkit.WebViewClient;

import androidx.annotation.RequiresApi;

import com.nintersoft.bibliotecaufabc.R;
import com.nintersoft.bibliotecaufabc.book_renewal_model.BookRenewalDAO;
import com.nintersoft.bibliotecaufabc.book_renewal_model.BookRenewalDatabaseSingletonFactory;
import com.nintersoft.bibliotecaufabc.book_renewal_model.BookRenewalProperties;
import com.nintersoft.bibliotecaufabc.synchronization.SyncService;
import com.nintersoft.bibliotecaufabc.utilities.GlobalConstants;
import com.nintersoft.bibliotecaufabc.utilities.GlobalFunctions;
import com.nintersoft.bibliotecaufabc.utilities.GlobalVariables;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;

public class SyncWebClient extends WebViewClient {
    private int login_page_finished;
    private int login_home_finished;
    private int renewal_page_finished;

    private String user_login;
    private String user_password;

    private Context mContext;

    public SyncWebClient(Context context){
        super();
        mContext = context;

        login_page_finished = 0;
        login_home_finished = 0;
        renewal_page_finished = 0;

        loadSettings();
    }

    private void loadSettings(){
        SharedPreferences pref = PreferenceManager.getDefaultSharedPreferences(mContext);
        GlobalVariables.storeUserFormData = pref.getBoolean(mContext.getString(R.string.key_privacy_store_password), true);

        user_login = pref.getString(mContext.getString(R.string.key_privacy_login_username), "");
        user_password = pref.getString(mContext.getString(R.string.key_privacy_login_password), "");

        if (user_login == null || user_password == null)
            ((SyncService)mContext).finish();
        if (!GlobalVariables.storeUserFormData || user_login.isEmpty()
                || user_password.isEmpty())
            ((SyncService)mContext).finish();
    }

    @Override
    public boolean shouldOverrideUrlLoading(WebView view, WebResourceRequest request) {
        return false;
    }

    @Override
    public boolean shouldOverrideUrlLoading(WebView view, String url) {
        return false;
    }

    @Override
    public void onPageFinished(final WebView view, String url) {
        super.onPageFinished(view, url);

        if (url.contains(GlobalConstants.URL_ACCESS_PAGE)){
            if (login_page_finished == 0){
                String script = String.format("%1$s \nperformLogin(\"%2$s\",\"%3$s\");",
                        GlobalFunctions.getScriptFromAssets(mContext, "javascript/login_scraper.js"),
                        user_login, user_password);
                view.evaluateJavascript(script, null);
                login_page_finished++;

                ((SyncService)mContext).killLater();
            }

            String script = String.format("%1$s \ncheckForErrors();",
                    GlobalFunctions.getScriptFromAssets(mContext, "javascript/login_scraper.js"));
            view.evaluateJavascript(script, new ValueCallback<String>() {
                @Override
                public void onReceiveValue(final String value) {
                    try {
                        JSONObject result = new JSONObject(value);
                        if (result.getBoolean("hasFormError")) {
                            GlobalFunctions.createSyncNotification(mContext,
                                    R.string.notification_sync_error_title,
                                    R.string.notification_sync_error_message,
                                    GlobalConstants.SYNC_NOTIFICATION_UPDATE_ID);
                            ((SyncService)mContext).finish();
                        }
                    } catch (JSONException e){
                        ((SyncService)mContext).retryAndFinish();
                    }
                }
            });
        }
        else if (url.contains(GlobalConstants.URL_LIBRARY_HOME)){
            if (login_home_finished < 1){
                login_home_finished++;
                return;
            }

            renewal_page_finished = 0;

            view.loadUrl(GlobalConstants.URL_LIBRARY_RENEWAL);
        }
        else if (url.contains(GlobalConstants.URL_LIBRARY_RENEWAL)){
            if (renewal_page_finished < 1){
                renewal_page_finished++;
                return;
            }

            login_page_finished = 0;
            login_home_finished = 0;

            String script = String.format("%1$s \ngetRenewals();",
                    GlobalFunctions.getScriptFromAssets(mContext, "javascript/renewal_scraper.js"));
            view.evaluateJavascript(script, new ValueCallback<String>() {
                @Override
                public void onReceiveValue(final String value) {
                    try{
                        JSONObject result = new JSONObject(value);
                        if (result.getBoolean("connected"))
                            setReservationBooks(result.getBoolean("hasErrorDiv") ? new JSONArray() :
                                    result.getJSONArray("renewalBooks"));
                        else view.loadUrl(GlobalConstants.URL_LIBRARY_LOGIN);
                    } catch (JSONException e){
                        ((SyncService)mContext).retryAndFinish();
                    }
                }
            });
        }
    }

    @RequiresApi(api = Build.VERSION_CODES.M)
    @Override
    public void onReceivedError(WebView view, WebResourceRequest request, WebResourceError error) {
        super.onReceivedError(view, request, error);
        ((SyncService)mContext).retryAndFinish();
    }

    @Override
    public void onReceivedError(WebView view, int errorCode, String description, String failingUrl) {
        super.onReceivedError(view, errorCode, description, failingUrl);
        ((SyncService)mContext).retryAndFinish();
    }

    private void setReservationBooks(JSONArray jsResultsArr){
        ArrayList<BookRenewalProperties> availableBooks = new ArrayList<>();

        try {
            availableBooks.clear();
            for (int i = 0; i < jsResultsArr.length(); ++i){
                JSONObject jsBook = jsResultsArr.getJSONObject(i);

                BookRenewalProperties newBook = new BookRenewalProperties();
                newBook.setId(i);
                newBook.setTitle(jsBook.getString("title"));
                newBook.setLibrary(jsBook.getString("library"));
                newBook.setPatrimony(jsBook.getString("patrimony"));
                newBook.setDate(jsBook.getString("date"));
                newBook.setRenewalLink(jsBook.getString("renewal_link"));

                availableBooks.add(newBook);
            }
            bindAlarms(availableBooks);
            //_DEBUG: Remove function call and scope
            GlobalFunctions.writeToFile(jsResultsArr.toString(), "complete");
            ((SyncService)mContext).finish();
        }catch (JSONException e){
            ((SyncService)mContext).retryAndFinish();
        }
    }

    private void bindAlarms(ArrayList<BookRenewalProperties> availableBooks){
        BookRenewalDAO dao = BookRenewalDatabaseSingletonFactory.getInstance().bookRenewalDAO();
        GlobalFunctions.cancelExistingScheduledAlarms(mContext, dao);

        dao.removeAll();
        for (BookRenewalProperties b: availableBooks) dao.insert(b);

        GlobalFunctions.scheduleRenewalAlarms(mContext, dao);
        GlobalFunctions.scheduleSyncNotification(mContext, 432000000);

        SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(mContext);
        SharedPreferences.Editor editor = preferences.edit();
        editor.putLong(mContext.getString(R.string.key_synchronization_schedule), System.currentTimeMillis());
        editor.apply();
    }
}

File: BibliotecaUFABC/app/src/main/java/com/nintersoft/bibliotecaufabc/webclient/SyncWebClient.kt
package com.nintersoft.bibliotecaufabc.webclient

import android.content.Context
import android.os.Build
import android.webkit.*
import androidx.annotation.RequiresApi
import androidx.preference.PreferenceManager
import com.nintersoft.bibliotecaufabc.R
import com.nintersoft.bibliotecaufabc.global.Constants
import com.nintersoft.bibliotecaufabc.global.Functions
import com.nintersoft.bibliotecaufabc.model.AppDatabase
import com.nintersoft.bibliotecaufabc.model.renewal.BookRenewal
import com.nintersoft.bibliotecaufabc.synchronization.SyncService
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import java.text.SimpleDateFormat
import java.util.*

class SyncWebClient(private val context : Context)
    : WebViewClient() {

    private var loginPageFinished = 0
    private var loginServicesFinished = 0
    private var renewalPageFinished = 0

    private var uLogin : String? = null
    private var uPassword : String? = null

    init {
        PreferenceManager.getDefaultSharedPreferences(context).let {
            uLogin = it.getString(context.
                getString(R.string.key_privacy_login_username), "")
            uPassword = it.getString(context.
                getString(R.string.key_privacy_login_password), "")

            if (!it.getBoolean(context
                    .getString(R.string.key_privacy_store_password), true) ||
                    uLogin.isNullOrEmpty() || uPassword.isNullOrEmpty())
                (context as SyncService).setStatus(SyncService.Companion.LStatus.FINISHED_SUCCESS)
        }
    }

    override fun shouldOverrideUrlLoading(view: WebView?, request: WebResourceRequest?): Boolean {
        return false
    }

    override fun shouldOverrideUrlLoading(view: WebView?, url: String?): Boolean { return false }

    override fun onPageFinished(view: WebView?, url: String?) {
        super.onPageFinished(view, url)
        if (url.isNullOrEmpty()) return

        when {
            url.contains(Constants.URL_LIBRARY_LOGIN_P) -> {
                if (loginPageFinished == 0) {
                    loginPageFinished++
                    return
                }

                renewalPageFinished = 0
                loginServicesFinished = 0
                val script = "${Functions.scriptFromAssets("js/login_scp.js")}\nperformLogin('$uLogin','$uPassword');"
                view?.evaluateJavascript(script, null)

                with (context as SyncService) {
                    scheduleChecking()
                    killLater()
                }
            }
            url.contains(Constants.URL_LIBRARY_SERVICES) -> {
                if (loginServicesFinished < 1){
                    loginServicesFinished++
                    return
                }

                renewalPageFinished = 0
                view?.loadUrl(Constants.URL_LIBRARY_RENEWAL)
            }
            url.contains(Constants.URL_LIBRARY_RENEWAL) -> {
                if (renewalPageFinished < 1){
                    renewalPageFinished++
                    return
                }

                loginPageFinished = 0
                loginServicesFinished = 0

                val script = "${Functions.scriptFromAssets("js/renewal_scp.js")}\ngetRenewals();"
                view?.evaluateJavascript(script, ({result ->
                    try {
                        with (JSONObject(result)) {
                            if (getBoolean("connected"))
                                setReservationBooks(if (getBoolean("hasErrorDiv"))
                                    JSONArray() else getJSONArray("renewalBooks"))
                            else view.loadUrl(Constants.URL_LIBRARY_LOGIN_P)
                        }
                    } catch (_ : JSONException) { (context as SyncService).setStatus(SyncService.Companion.LStatus.FINISHED_FAILURE) }
                }))
            }
        }
    }

    @RequiresApi(api = Build.VERSION_CODES.M)
    override fun onReceivedError(view: WebView?, request: WebResourceRequest?,
        error: WebResourceError?) {
        super.onReceivedError(view, request, error)
        (context as SyncService).setStatus(SyncService.Companion.LStatus.FINISHED_FAILURE)
    }

    @Suppress("DEPRECATION")
    override fun onReceivedError(view: WebView?, errorCode: Int, description: String?,
                                 failingUrl: String?) {
        super.onReceivedError(view, errorCode, description, failingUrl)
        (context as SyncService).setStatus(SyncService.Companion.LStatus.FINISHED_FAILURE)
    }

    private fun setReservationBooks(jsBooks : JSONArray){
        try {
            bindAlarms(arrayListOf<BookRenewal>().also {list ->
                for (i in 0 until jsBooks.length()){
                    list.add(BookRenewal().also {book ->
                        with (jsBooks.getJSONObject(i)){
                            book.id = i.toLong()
                            book.title = getString("title")
                            book.library = getString("library")
                            book.patrimony = getString("patrimony")
                            book.date = SimpleDateFormat("dd/MM/yy", Locale.getDefault()).
                                parse(getString("date").split(" ")[2])
                            book.renewalLink = getString("renewal_link")
                        }
                    })
                }
            })
            (context as SyncService).setStatus(SyncService.Companion.LStatus.FINISHED_SUCCESS)
        } catch (_ : JSONException) { (context as SyncService).setStatus(SyncService.Companion.LStatus.FINISHED_FAILURE) }
    }

    private fun bindAlarms(books : List<BookRenewal>){
        AppDatabase.getInstance()?.bookRenewalDAO()?.let {dao ->
            GlobalScope.launch {
                dao.removeAll()
                books.forEach { dao.insert(it) }
            }
            Functions.scheduleRenewalAlarms()

            PreferenceManager.getDefaultSharedPreferences(context).edit()
                .putLong(context.getString(R.string.key_synchronization_schedule),
                    System.currentTimeMillis()).apply()
        }

    }
}